In this practice code, we are going to use Ozone dataset. Dictionary of this dataset can be found in https://stat.ethz.ch/R-manual/R-devel/library/datasets/html/airquality.html

In [1]:
import pandas as pd
import numpy as np

In [2]:
url = "https://raw.githubusercontent.com/ga-students/SF-DAT-20/master/Data/ozone.csv"
OzoneData = pd.read_csv(url)

#### Explore the dataset and decide which variables suffer from missing data

In [3]:
print(len(OzoneData))
OzoneData.describe()

153


Ozone     Solar.R        Wind        Temp       Month         Day
count  116.000000  146.000000  153.000000  153.000000  153.000000  153.000000
mean    42.129310  185.931507    9.957516   77.882353    6.993464   15.803922
std     32.987885   90.058422    3.523001    9.465270    1.416522    8.864520
min      1.000000    7.000000    1.700000   56.000000    5.000000    1.000000
25%     18.000000  115.750000    7.400000   72.000000    6.000000    8.000000
50%     31.500000  205.000000    9.700000   79.000000    7.000000   16.000000
75%     63.250000  258.750000   11.500000   85.000000    8.000000   23.000000
max    168.000000  334.000000   20.700000   97.000000    9.000000   31.000000

Answer: It seems like Ozone and Solar.R suffer from missing values

#### Let's drop rows that have missing values in all the columns you indentified above

Hint: in dropna() if you set how = 'all', it will only drop columns that are suffering from missing values at all varaibles you introduce in subset. If you want to get rid of the row that contains missing values in any of the variables you specify, then you shall set how = 'any'

df.dropna(how = 'all',subset = ['Var1','Var2','Var3'],inplace = True)

The above code will check if all 3 variables specified in df have missing values, if they all have missing values it will drop that row.

In [4]:
OzoneData.dropna(how = 'all', subset = ['Ozone', 'Solar.R'], inplace = True)


In [5]:
OzoneData.describe()
# it seems like 2 observations are dropped

Ozone     Solar.R        Wind        Temp       Month         Day
count  116.000000  146.000000  151.000000  151.000000  151.000000  151.000000
mean    42.129310  185.931507    9.941722   78.165563    7.019868   15.801325
std     32.987885   90.058422    3.524984    9.198138    1.406984    8.832531
min      1.000000    7.000000    1.700000   57.000000    5.000000    1.000000
25%     18.000000  115.750000    7.400000   73.000000    6.000000    8.000000
50%     31.500000  205.000000    9.700000   79.000000    7.000000   16.000000
75%     63.250000  258.750000   11.500000   85.000000    8.000000   23.000000
max    168.000000  334.000000   20.700000   97.000000    9.000000   31.000000

#### Spoiler! If everything is going according to plan you shall be left by 151 observations. Also, it seemed like the first two varibales had missing values. Now plese make a copy of your dataframe into a dataframe named OzoneImputeMean. Also, please use mean of the variables to fill in missing values in OzoneImputeMean

In [6]:
OzoneImputeMean = OzoneData
OzoneImputeMean['Ozone'].fillna(value = np.mean(OzoneImputeMean['Ozone']), inplace = True)
OzoneImputeMean['Solar.R'].fillna(value = np.mean(OzoneImputeMean['Solar.R']), inplace = True)


In [7]:
#Check if missing values are filled
OzoneImputeMean.describe()

#yes everything seems to be fine

Ozone     Solar.R        Wind        Temp       Month         Day
count  151.000000  151.000000  151.000000  151.000000  151.000000  151.000000
mean    42.129310  185.931507    9.941722   78.165563    7.019868   15.801325
std     28.884028   88.544727    3.524984    9.198138    1.406984    8.832531
min      1.000000    7.000000    1.700000   57.000000    5.000000    1.000000
25%     21.000000  119.000000    7.400000   73.000000    6.000000    8.000000
50%     42.129310  197.000000    9.700000   79.000000    7.000000   16.000000
75%     46.500000  257.000000   11.500000   85.000000    8.000000   23.000000
max    168.000000  334.000000   20.700000   97.000000    9.000000   31.000000

#### Now it's time for imputing using linear regression lines

In [10]:
#### Before we start let's define dummy variables for variable Month - don't worry about day!
MonthDummy = pd.get_dummies(OzoneData.Month, prefix = 'Month')
del MonthDummy['Month_9']
OzoneData = pd.concat([OzoneData, MonthDummy], axis=1)
OzoneData.head()

Ozone     Solar.R  Wind  Temp  Month  Day  Month_5  Month_6  Month_7  \
0     41  190.000000   7.4    67      5    1        1        0        0   
1     36  118.000000   8.0    72      5    2        1        0        0   
2     12  149.000000  12.6    74      5    3        1        0        0   
3     18  313.000000  11.5    62      5    4        1        0        0   
5     28  185.931507  14.9    66      5    6        1        0        0   

   Month_8  
0        0  
1        0  
2        0  
3        0  
5        0

In [11]:
#now let's explore correlation Matrix
OzoneData.corr()

Ozone   Solar.R      Wind      Temp     Month       Day   Month_5  \
Ozone    1.000000  0.302970 -0.534163  0.630583  0.151089 -0.011472 -0.281111   
Solar.R  0.302970  1.000000 -0.055581  0.271990 -0.073886 -0.147120 -0.023842   
Wind    -0.534163 -0.055581  1.000000 -0.466032 -0.175317  0.042365  0.237781   
Temp     0.630583  0.271990 -0.466032  1.000000  0.397427 -0.136876 -0.637816   
Month    0.151089 -0.073886 -0.175317  0.397427  1.000000 -0.007727 -0.702257   
Day     -0.011472 -0.147120  0.042365 -0.136876 -0.007727  1.000000  0.011003   
Month_5 -0.281111 -0.023842  0.237781 -0.637816 -0.702257  0.011003  1.000000   
Month_6 -0.065820  0.023896  0.046054  0.050753 -0.362131 -0.017044 -0.242766   
Month_7  0.251524  0.175960 -0.144638  0.318103 -0.007201  0.011471 -0.247805   
Month_8  0.264054 -0.073214 -0.166105  0.321680  0.355246  0.011471 -0.247805   

          Month_6   Month_7   Month_8  
Ozone   -0.065820  0.251524  0.264054  
Solar.R  0.023896  0.175960 -0.073214  
Wind     0.046054 -0.144638 -0.166105  
Temp     0.050753  0.318103  0.321680  
Month   -0.362131 -0.007201  0.355246  
Day     -0.017044  0.011471  0.011471  
Month_5 -0.242766 -0.247805 -0.247805  
Month_6  1.000000 -0.253080 -0.253080  
Month_7 -0.253080  1.000000 -0.258333  
Month_8 -0.253080 -0.258333  1.000000

#### What seems to be the list of best variables can define Ozone? how about Solar.R?

Answer: it seems like Solar.R, Wind and Temp, and month dummies are good for predicting Ozone.
        For Solar.R, it seems like Ozone, Temp are the most important variables

In [12]:
# now let's use a regression model to predict Ozone. First drop NaN values in Ozone and save it in OzoneDroppedValues_Ozone
# the run a regression line on variables of interest and check significancy of your model
# if in a multi class dummy variable case, you see a few of the dummy variables are not significant but the rest are, 
#you shall either drop all or keep all. Otherwise, selecting the base dummy will become important
# use these variables ['Solar.R','Wind','Temp','Month_5','Month_6','Month_7','Month_8'] to predict
from sklearn.linear_model import LinearRegression
from sklearn import feature_selection

linreg = LinearRegression()
OzoneDroppedValues_Ozone = OzoneData.dropna(subset = ['Ozone'])
X = OzoneDroppedValues_Ozone[['Solar.R','Wind','Temp','Month_5','Month_6','Month_7','Month_8']]
y = OzoneDroppedValues_Ozone['Ozone']
linreg.fit(X,y)
pvals = feature_selection.f_regression(X,y)[1]
print(pvals)
#Almost everything is significant - 

[  1.56023855e-04   1.62506497e-12   4.07521160e-18   4.71539264e-04
   4.21995500e-01   1.83732627e-03   1.05239684e-03]


In [13]:
# now fill in null values of OzoneData['Ozone'] by predicted values
X = OzoneData[['Solar.R','Wind','Temp','Month_5','Month_6','Month_7','Month_8']]
y = OzoneData['Ozone']
OzoneData['Ozone_predicted'] = linreg.predict(X)
OzoneData['Ozone'].fillna(value = OzoneData['Ozone_predicted'], inplace = True)

In [14]:
# Now repeat previous steps for Solar.R variable using ['Ozone','Wind','Temp]
OzoneDroppedValues_SolarR = OzoneData.dropna(subset = ['Solar.R'])
X = OzoneDroppedValues_SolarR[['Ozone','Wind','Temp']]
y = OzoneDroppedValues_SolarR['Solar.R']
linreg.fit(X,y)
pvals = feature_selection.f_regression(X,y)[1]
print(pvals)
#Check p-values of your model - if a variable is not significant, drop it

### We shall drop Wind and rerun our regressions again
X = OzoneDroppedValues_SolarR[['Ozone','Temp']]
y = OzoneDroppedValues_SolarR['Solar.R']
linreg.fit(X,y)
pvals = feature_selection.f_regression(X,y)[1]
print(pvals)

[  1.56023855e-04   4.97876020e-01   7.29051873e-04]
[ 0.00015602  0.00072905]


In [15]:
# now fill in null values of OzoneData['Solar.R'] by predicted values
X = OzoneData[['Ozone','Temp']]
y = OzoneData['Solar.R']
OzoneData['Solar.R_predicted'] = linreg.predict(X)
OzoneData['Solar.R'].fillna(value = OzoneData['Solar.R_predicted'], inplace = True)

#### Now check your filled data - if your predicted values are more than maximum or less than minimum, replace them by max and min

In [14]:
OzoneData.describe()
OzoneData.head()

Ozone     Solar.R        Wind        Temp       Month         Day  \
count  151.000000  151.000000  151.000000  151.000000  151.000000  151.000000   
mean    42.129310  185.931507    9.941722   78.165563    7.019868   15.801325   
std     28.884028   88.544727    3.524984    9.198138    1.406984    8.832531   
min      1.000000    7.000000    1.700000   57.000000    5.000000    1.000000   
25%     21.000000  119.000000    7.400000   73.000000    6.000000    8.000000   
50%     42.129310  197.000000    9.700000   79.000000    7.000000   16.000000   
75%     46.500000  257.000000   11.500000   85.000000    8.000000   23.000000   
max    168.000000  334.000000   20.700000   97.000000    9.000000   31.000000   

          Month_5     Month_6     Month_7     Month_8  Ozone_predicted  \
count  151.000000  151.000000  151.000000  151.000000       151.000000   
mean     0.192053    0.198675    0.205298    0.205298        42.129310   
std      0.395225    0.400331    0.405263    0.405263        20.928089   
min      0.000000    0.000000    0.000000    0.000000       -11.868224   
25%      0.000000    0.000000    0.000000    0.000000        29.652699   
50%      0.000000    0.000000    0.000000    0.000000        43.308777   
75%      0.000000    0.000000    0.000000    0.000000        55.316057   
max      1.000000    1.000000    1.000000    1.000000        90.567997   

       Solar.R_predicted  
count         151.000000  
mean          185.931507  
std            28.371236  
min           133.624407  
25%           164.266189  
50%           183.130275  
75%           200.751117  
max           273.806250

In [ ]:
### Fortunately, everythin is within acceptable range!

## Bonus: In original question, replace missing values of Ozone and Solar.R by Single Imputation Regression Line with Error

In [ ]:
url = "https://raw.githubusercontent.com/ga-students/SF-DAT-20/master/Data/ozone.csv"
OzoneData = pd.read_csv(url)
OzoneData.dropna(how = 'all', subset = ['Ozone', 'Solar.R'], inplace = True)





